src/events/tls.py

```diff
@@ -0,0 +1,163 @@
+#!/usr/bin/env python3
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Event handler for related applications on the `certificates` relation interface."""
+import base64
+import logging
+import re
+from typing import TYPE_CHECKING
+
+from charms.tls_certificates_interface.v3.tls_certificates import (
+    CertificateAvailableEvent,
+    TLSCertificatesRequiresV3,
+    generate_csr,
+    generate_private_key,
+)
+from ops.charm import ActionEvent, RelationCreatedEvent, RelationJoinedEvent
+from ops.framework import EventBase, Object
+
+if TYPE_CHECKING:
+    from charm import OpensearchDasboardsCharm
+
+logger = logging.getLogger(__name__)
+
+
+class TLSEvents(Object):
+    """Event handlers for related applications on the `certificates` relation interface."""
+
+    def __init__(self, charm):
+        super().__init__(charm, "tls")
+        self.charm: "OpensearchDasboardsCharm" = charm
+        self.certificates = TLSCertificatesRequiresV3(self.charm, "certificates")
+
+        self.framework.observe(
+            getattr(self.charm.on, "certificates_relation_created"),
+            self._on_certs_relation_created,
+        )
+        self.framework.observe(
+            getattr(self.charm.on, "certificates_relation_joined"), self._on_certs_relation_joined
+        )
+        self.framework.observe(
+            getattr(self.certificates.on, "certificate_available"), self._on_certificate_available
+        )
+        self.framework.observe(
+            getattr(self.certificates.on, "certificate_expiring"), self._on_certificate_expiring
+        )
+        self.framework.observe(
+            getattr(self.charm.on, "certificates_relation_broken"), self._on_certs_relation_broken
+        )
+
+        self.framework.observe(
+            getattr(self.charm.on, "set_tls_private_key_action"), self._set_tls_private_key
+        )
+
+    def _on_certs_relation_created(self, event: RelationCreatedEvent) -> None:
+        """Handler for `certificates_relation_created` event."""
+        if not self.charm.unit.is_leader():
+            return
+
+        self.charm.state.cluster.update({"tls": "enabled", "switching-encryption": "started"})
+
+    def _on_certs_relation_joined(self, event: RelationJoinedEvent) -> None:
```

Comment:
I recommend merging `_created` and `_joined`. We had a discussion with the team behind it and eventually we should all be using `_created` instead: https://github.com/canonical/tls-certificates-interface/issues/144

The `-joined` event happens on a "per-new-unit joining" basis, whereas "-created"  happens once, when the relation has been created.

Reply:
I see your point. TBH here `created` is really not doing much, except reporting on the progress.
(We may not need this really, I'm adding a reveiw on this point to the ['cleanup ticket'](https://warthogs.atlassian.net/browse/DPE-3954)). If you can accept, then I'd close this thread now.